[Automation] native-image run fails for org.htrace:htrace-core:3.0.4

## native-image run on ubuntu-latest / JDK 25 / mode current-defaults

Failure kind: native-image run
OS: ubuntu-latest
JDK: 25
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.0.4" ./gradlew clean test -Pcoordinates="org.htrace:htrace-core:3.0.4"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25727047580/job/75542581547
Last 300 lines of the log:
```console
[5/8] Inlining methods...     [***]                                                                      (2.1s @ 0.65GB)
[6/8] Compiling methods...    [****]                                                                    (18.4s @ 0.79GB)
[7/8] Laying out methods...   [**]                                                                       (3.4s @ 0.75GB)
[8/8] Creating image...       [**]                                                                       (3.1s @ 0.89GB)
   9.48MB (35.74%) for code area:    22,517 compilation units
  13.37MB (50.38%) for image heap:  211,568 objects and 153 resources
   3.68MB (13.89%) for other data
  26.54MB in total image size, 24.51MB in total file size
------------------------------------------------------------------------------------------------------------------------
Top 10 origins of code area:                                Top 10 object types in image heap:
   4.57MB java.base                                            2.76MB byte[] for code metadata
   2.09MB svm.jar (Native Image)                               2.59MB byte[] for java.lang.String
 588.50kB assertj-core-3.22.0.jar                              1.36MB java.lang.String
 481.23kB java.xml                                             1.04MB java.lang.Class
 209.19kB junit-jupiter-engine-5.9.2.jar                     864.55kB com.oracle.svm.core.hub.DynamicHubCompanion
 121.80kB junit-platform-launcher-1.9.2.jar                  707.04kB byte[] for general heap data
 111.27kB commons-logging-1.1.1.jar                          450.40kB java.util.concurrent.ConcurrentHashMap$Node
 104.38kB java.logging                                       381.38kB heap alignment
 100.10kB junit-platform-engine-1.9.2.jar                    300.93kB java.util.HashMap$Node
  81.79kB junit-4.13.2.jar                                   296.57kB byte[] for reflection metadata
 673.35kB for 28 more packages                                 2.62MB for 2035 more object types
                            Use '--emit build-report' to create a report with more details.
------------------------------------------------------------------------------------------------------------------------
Security report:
 - Binary includes Java deserialization.
 - CycloneDX SBOM with 23 component(s) is embedded in binary (1.16kB). 3 type(s) could not be associated to a component.
 - Advanced obfuscation not enabled; enable with '-H:AdvancedObfuscation=""' (experimental support).
------------------------------------------------------------------------------------------------------------------------
Recommendations:
 G1GC: Use the G1 GC ('--gc=G1') for improved latency and throughput.
 PGO:  Use Profile-Guided Optimizations ('--pgo') for improved throughput.
 FUTR: Use '--future-defaults=all' to prepare for future releases.
 HEAP: Set max heap for improved and more predictable memory usage.
 CPU:  Enable more CPU features with '-march=native' for improved performance.
------------------------------------------------------------------------------------------------------------------------
                        5.3s (6.9% of total time) in 773 GCs | Peak RSS: 1.71GB | CPU load: 3.63
------------------------------------------------------------------------------------------------------------------------
Build artifacts:
 /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.htrace/htrace-core/3.0.4/build/native/nativeTestCompile/libmanagement_ext.so (jdk_library)
 /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.htrace/htrace-core/3.0.4/build/native/nativeTestCompile/org.htrace.htrace-core_tests-tests (executable)
 /tmp/native-image-build-output.json (build_info)
========================================================================================================================
Finished generating 'org.htrace.htrace-core_tests-tests' in 1m 16s.
[native-image-plugin] Native Image written to: /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.htrace/htrace-core/3.0.4/build/native/nativeTestCompile

> Task :nativeTest FAILED
[junit-platform-native] WARNING: -djunit.platform.listeners.uid.tracking.output.dir not specified, trying to find test-ids on default Gradle/Maven locations. As this is a fallback mode, it could take a while. This should only happen if you are running tests executable manually and you didn't pass uid output directory with -djunit.platform.listeners.uid.tracking.output.dir=<path>.
[junit-platform-native] WARNING: Using test-ids from default Gradle project location:/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.htrace/htrace-core/3.0.4/build/test-results/test/testlist
[junit-platform-native] Running in 'test listener' mode using files matching pattern [junit-platform-unique-ids*] found in folder [/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.htrace/htrace-core/3.0.4/build/test-results/test/testlist] and its subfolders.
JUnit Platform on Native Image - report
----------------------------------------

org_htrace.htrace_core.Htrace_coreTest > traceInfoCanStartDetachedSpanAndContinueItLater() FAILED

org_htrace.htrace_core.Htrace_coreTest > traceProxyCreatesSpansAroundInterfaceMethodCalls() FAILED

org_htrace.htrace_core.Htrace_coreTest > configurationParsesDefaultsAndSamplersMakeTracingDecisions() FAILED

org_htrace.htrace_core.Htrace_coreTest > callableRunnableAndExecutorWrappersPropagateTheCurrentSpan() FAILED

org_htrace.htrace_core.Htrace_coreTest > nestedScopesDeliverAnnotatedSpansAndBuildTraceTree() FAILED

org_htrace.htrace_core.Htrace_coreTest > traceInfoSamplerCanAcceptOrRejectRemoteTraceContinuation() FAILED

org_htrace.htrace_core.Htrace_coreTest > nullScopeAndStoppedMilliSpanExposeStableValues() FAILED

org_htrace.htrace_core.Htrace_coreTest > samplerContextControlsRootSpanCreation() FAILED

org_htrace.htrace_core.Htrace_coreTest > localFileReceiverWritesSpanJson() FAILED


Failures (9):
  JUnit Jupiter:Htrace_coreTest:traceInfoCanStartDetachedSpanAndContinueItLater()
    MethodSource [className = 'org_htrace.htrace_core.Htrace_coreTest', methodName = 'traceInfoCanStartDetachedSpanAndContinueItLater', methodParameterTypes = '']
    => org.graalvm.nativeimage.MissingReflectionRegistrationError: Cannot reflectively invoke method 'public java.lang.ClassLoader java.lang.Thread.getContextClassLoader()'. To allow this operation, add the following to the 'reflection' section of 'reachability-metadata.json' and rebuild the native image:

  {
    "type": "java.lang.Thread",
    "methods": [
      {
        "name": "getContextClassLoader",
        "parameterTypes": []
      }
    ]
  }

The 'reachability-metadata.json' file should be located in 'META-INF/native-image/<group-id>/<artifact-id>/' of your project. For further help, see https://www.graalvm.org/latest/reference-manual/native-image/metadata/#reflection
       org.graalvm.nativeimage.builder/com.oracle.svm.core.reflect.MissingReflectionRegistrationUtils.reportInvokedExecutable(MissingReflectionRegistrationUtils.java:110)
       java.base@25.0.3/java.lang.reflect.Method.acquireMethodAccessor(Method.java:149)
       java.base@25.0.3/java.lang.reflect.Method.invoke(Method.java:562)
       org.apache.commons.logging.LogFactory.directGetContextClassLoader(LogFactory.java:901)
       org.apache.commons.logging.LogFactory$1.run(LogFactory.java:862)
       java.base@25.0.3/java.security.AccessController.doPrivileged(AccessController.java:74)
       org.apache.commons.logging.LogFactory.getContextClassLoaderInternal(LogFactory.java:859)
       org.apache.commons.logging.LogFactory.getFactory(LogFactory.java:423)
       org.apache.commons.logging.LogFactory.getLog(LogFactory.java:685)
       org.htrace.Tracer.<clinit>(Tracer.java:33)
       [...]
  JUnit Jupiter:Htrace_coreTest:traceProxyCreatesSpansAroundInterfaceMethodCalls()
    MethodSource [className = 'org_htrace.htrace_core.Htrace_coreTest', methodName = 'traceProxyCreatesSpansAroundInterfaceMethodCalls', methodParameterTypes = '']
    => java.lang.NoClassDefFoundError: Could not initialize class org.htrace.Tracer
       org.htrace.Trace.setProcessId(Trace.java:106)
       org_htrace.htrace_core.Htrace_coreTest.traceProxyCreatesSpansAroundInterfaceMethodCalls(Htrace_coreTest.java:307)
       java.base@25.0.3/java.lang.reflect.Method.invoke(Method.java:565)
       org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
       org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
       org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
       org.junit.jupiter.api.AssertTimeoutPreemptively.lambda$submitTask$3(AssertTimeoutPreemptively.java:95)
       java.base@25.0.3/java.util.concurrent.FutureTask.run(FutureTask.java:328)
       java.base@25.0.3/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
       java.base@25.0.3/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
       [...]
  JUnit Jupiter:Htrace_coreTest:configurationParsesDefaultsAndSamplersMakeTracingDecisions()
    MethodSource [className = 'org_htrace.htrace_core.Htrace_coreTest', methodName = 'configurationParsesDefaultsAndSamplersMakeTracingDecisions', methodParameterTypes = '']
    => org.graalvm.nativeimage.MissingReflectionRegistrationError: Cannot reflectively invoke method 'public java.lang.ClassLoader java.lang.Thread.getContextClassLoader()'. To allow this operation, add the following to the 'reflection' section of 'reachability-metadata.json' and rebuild the native image:

  {
    "type": "java.lang.Thread",
    "methods": [
      {
        "name": "getContextClassLoader",
        "parameterTypes": []
      }
    ]
  }

The 'reachability-metadata.json' file should be located in 'META-INF/native-image/<group-id>/<artifact-id>/' of your project. For further help, see https://www.graalvm.org/latest/reference-manual/native-image/metadata/#reflection
       org.graalvm.nativeimage.builder/com.oracle.svm.core.reflect.MissingReflectionRegistrationUtils.reportInvokedExecutable(MissingReflectionRegistrationUtils.java:110)
       java.base@25.0.3/java.lang.reflect.Method.acquireMethodAccessor(Method.java:149)
       java.base@25.0.3/java.lang.reflect.Method.invoke(Method.java:562)
       org.apache.commons.logging.LogFactory.directGetContextClassLoader(LogFactory.java:901)
       org.apache.commons.logging.LogFactory$1.run(LogFactory.java:862)
       java.base@25.0.3/java.security.AccessController.doPrivileged(AccessController.java:74)
       org.apache.commons.logging.LogFactory.getContextClassLoaderInternal(LogFactory.java:859)
       org.apache.commons.logging.LogFactory.getFactory(LogFactory.java:423)
       org.apache.commons.logging.LogFactory.getLog(LogFactory.java:685)
       org.htrace.HTraceConfiguration.<clinit>(HTraceConfiguration.java:31)
       [...]
  JUnit Jupiter:Htrace_coreTest:callableRunnableAndExecutorWrappersPropagateTheCurrentSpan()
    MethodSource [className = 'org_htrace.htrace_core.Htrace_coreTest', methodName = 'callableRunnableAndExecutorWrappersPropagateTheCurrentSpan', methodParameterTypes = '']
    => java.lang.NoClassDefFoundError: Could not initialize class org.htrace.Tracer
       org.htrace.Trace.setProcessId(Trace.java:106)
       org_htrace.htrace_core.Htrace_coreTest.callableRunnableAndExecutorWrappersPropagateTheCurrentSpan(Htrace_coreTest.java:248)
       java.base@25.0.3/java.lang.reflect.Method.invoke(Method.java:565)
       org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
       org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
       org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
       org.junit.jupiter.api.AssertTimeoutPreemptively.lambda$submitTask$3(AssertTimeoutPreemptively.java:95)
       java.base@25.0.3/java.util.concurrent.FutureTask.run(FutureTask.java:328)
       java.base@25.0.3/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
       java.base@25.0.3/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
       [...]
  JUnit Jupiter:Htrace_coreTest:nestedScopesDeliverAnnotatedSpansAndBuildTraceTree()
    MethodSource [className = 'org_htrace.htrace_core.Htrace_coreTest', methodName = 'nestedScopesDeliverAnnotatedSpansAndBuildTraceTree', methodParameterTypes = '']
    => java.lang.NoClassDefFoundError: Could not initialize class org.htrace.Tracer
       org.htrace.Trace.setProcessId(Trace.java:106)
       org_htrace.htrace_core.Htrace_coreTest.nestedScopesDeliverAnnotatedSpansAndBuildTraceTree(Htrace_coreTest.java:82)
       java.base@25.0.3/java.lang.reflect.Method.invoke(Method.java:565)
       org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
       org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
       org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
       org.junit.jupiter.api.AssertTimeoutPreemptively.lambda$submitTask$3(AssertTimeoutPreemptively.java:95)
       java.base@25.0.3/java.util.concurrent.FutureTask.run(FutureTask.java:328)
       java.base@25.0.3/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
       java.base@25.0.3/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
       [...]
  JUnit Jupiter:Htrace_coreTest:traceInfoSamplerCanAcceptOrRejectRemoteTraceContinuation()
    MethodSource [className = 'org_htrace.htrace_core.Htrace_coreTest', methodName = 'traceInfoSamplerCanAcceptOrRejectRemoteTraceContinuation', methodParameterTypes = '']
    => java.lang.NoClassDefFoundError: Could not initialize class org.htrace.Tracer
       org.htrace.Trace.setProcessId(Trace.java:106)
       org_htrace.htrace_core.Htrace_coreTest.traceInfoSamplerCanAcceptOrRejectRemoteTraceContinuation(Htrace_coreTest.java:173)
       java.base@25.0.3/java.lang.reflect.Method.invoke(Method.java:565)
       org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
       org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
       org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
       org.junit.jupiter.api.AssertTimeoutPreemptively.lambda$submitTask$3(AssertTimeoutPreemptively.java:95)
       java.base@25.0.3/java.util.concurrent.FutureTask.run(FutureTask.java:328)
       java.base@25.0.3/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
       java.base@25.0.3/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
       [...]
  JUnit Jupiter:Htrace_coreTest:nullScopeAndStoppedMilliSpanExposeStableValues()
    MethodSource [className = 'org_htrace.htrace_core.Htrace_coreTest', methodName = 'nullScopeAndStoppedMilliSpanExposeStableValues', methodParameterTypes = '']
    => java.lang.NoClassDefFoundError: Could not initialize class org.htrace.Tracer
       org.htrace.Trace.isTracing(Trace.java:154)
       org.htrace.Trace.startSpan(Trace.java:84)
       org.htrace.Trace.startSpan(Trace.java:70)
       org_htrace.htrace_core.Htrace_coreTest.nullScopeAndStoppedMilliSpanExposeStableValues(Htrace_coreTest.java:337)
       java.base@25.0.3/java.lang.reflect.Method.invoke(Method.java:565)
       org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
       org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
       org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
       org.junit.jupiter.api.AssertTimeoutPreemptively.lambda$submitTask$3(AssertTimeoutPreemptively.java:95)
       java.base@25.0.3/java.util.concurrent.FutureTask.run(FutureTask.java:328)
       [...]
  JUnit Jupiter:Htrace_coreTest:samplerContextControlsRootSpanCreation()
    MethodSource [className = 'org_htrace.htrace_core.Htrace_coreTest', methodName = 'samplerContextControlsRootSpanCreation', methodParameterTypes = '']
    => java.lang.NoClassDefFoundError: Could not initialize class org.htrace.Tracer
       org.htrace.Trace.setProcessId(Trace.java:106)
       org_htrace.htrace_core.Htrace_coreTest.samplerContextControlsRootSpanCreation(Htrace_coreTest.java:138)
       java.base@25.0.3/java.lang.reflect.Method.invoke(Method.java:565)
       org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
       org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
       org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
       org.junit.jupiter.api.AssertTimeoutPreemptively.lambda$submitTask$3(AssertTimeoutPreemptively.java:95)
       java.base@25.0.3/java.util.concurrent.FutureTask.run(FutureTask.java:328)
       java.base@25.0.3/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
       java.base@25.0.3/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
       [...]
  JUnit Jupiter:Htrace_coreTest:localFileReceiverWritesSpanJson()
    MethodSource [className = 'org_htrace.htrace_core.Htrace_coreTest', methodName = 'localFileReceiverWritesSpanJson', methodParameterTypes = '']
    => org.graalvm.nativeimage.MissingReflectionRegistrationError: Cannot reflectively invoke method 'public java.lang.ClassLoader java.lang.Thread.getContextClassLoader()'. To allow this operation, add the following to the 'reflection' section of 'reachability-metadata.json' and rebuild the native image:

  {
    "type": "java.lang.Thread",
    "methods": [
      {
        "name": "getContextClassLoader",
        "parameterTypes": []
      }
    ]
  }

The 'reachability-metadata.json' file should be located in 'META-INF/native-image/<group-id>/<artifact-id>/' of your project. For further help, see https://www.graalvm.org/latest/reference-manual/native-image/metadata/#reflection
       org.graalvm.nativeimage.builder/com.oracle.svm.core.reflect.MissingReflectionRegistrationUtils.reportInvokedExecutable(MissingReflectionRegistrationUtils.java:110)
       java.base@25.0.3/java.lang.reflect.Method.acquireMethodAccessor(Method.java:149)
       java.base@25.0.3/java.lang.reflect.Method.invoke(Method.java:562)
       org.apache.commons.logging.LogFactory.directGetContextClassLoader(LogFactory.java:901)
       org.apache.commons.logging.LogFactory$1.run(LogFactory.java:862)
       java.base@25.0.3/java.security.AccessController.doPrivileged(AccessController.java:74)
       org.apache.commons.logging.LogFactory.getContextClassLoaderInternal(LogFactory.java:859)
       org.apache.commons.logging.LogFactory.getFactory(LogFactory.java:423)
       org.apache.commons.logging.LogFactory.getLog(LogFactory.java:685)
       org.htrace.impl.LocalFileSpanReceiver.<clinit>(LocalFileSpanReceiver.java:41)
       [...]

Test run finished after 8 ms
[         3 containers found      ]
[         0 containers skipped    ]
[         3 containers started    ]
[         0 containers aborted    ]
[         3 containers successful ]
[         0 containers failed     ]
[         9 tests found           ]
[         0 tests skipped         ]
[         9 tests started         ]
[         0 tests aborted         ]
[         0 tests successful      ]
[         9 tests failed          ]


[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.htrace/htrace-core/3.0.4/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
18 actionable tasks: 8 executed, 10 up-to-date

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':nativeTest'.
> Process 'command '/home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.htrace/htrace-core/3.0.4/build/native/nativeTestCompile/org.htrace.htrace-core_tests-tests'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

BUILD FAILED in 1m 19s

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for org.htrace:htrace-core:3.0.4 failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 1m 20s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 3.0.4 at 2026-05-12T10:36:21Z with exit code 1 after 81s
❌ ERROR: 'native-image run' for 3.0.4 failed with exit code 1.
FAILED[native-image run][3.0.4][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.0.4" ./gradlew clean test -Pcoordinates="org.htrace:htrace-core:3.0.4"]
```

--- a/metadata/org.htrace/htrace-core/3.0.4/reachability-metadata.json
+++ b/metadata/org.htrace/htrace-core/3.0.4/reachability-metadata.json
@@ -1,1 +1,184 @@
-{ }
+{
+  "reflection" : [
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.impl.LocalFileSpanReceiver$WriteSpanRunnable"
+      },
+      "type" : "java.lang.Object"
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.Tracer"
+      },
+      "type" : "java.lang.String"
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.HTraceConfiguration"
+      },
+      "type" : "java.lang.Thread",
+      "methods" : [
+        {
+          "name" : "getContextClassLoader",
+          "parameterTypes" : [ ]
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.TraceTree"
+      },
+      "type" : "java.lang.Thread",
+      "methods" : [
+        {
+          "name" : "getContextClassLoader",
+          "parameterTypes" : [ ]
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.Tracer"
+      },
+      "type" : "java.lang.Thread",
+      "methods" : [
+        {
+          "name" : "getContextClassLoader",
+          "parameterTypes" : [ ]
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.impl.LocalFileSpanReceiver"
+      },
+      "type" : "java.lang.Thread",
+      "methods" : [
+        {
+          "name" : "getContextClassLoader",
+          "parameterTypes" : [ ]
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.Tracer"
+      },
+      "type" : "org.apache.commons.logging.LogFactory"
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.HTraceConfiguration"
+      },
+      "type" : "org.apache.commons.logging.impl.Jdk14Logger",
+      "methods" : [
+        {
+          "name" : "<init>",
+          "parameterTypes" : [
+            "java.lang.String"
+          ]
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.Tracer"
+      },
+      "type" : "org.apache.commons.logging.impl.Jdk14Logger",
+      "methods" : [
+        {
+          "name" : "<init>",
+          "parameterTypes" : [
+            "java.lang.String"
+          ]
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.impl.LocalFileSpanReceiver"
+      },
+      "type" : "org.apache.commons.logging.impl.Jdk14Logger",
+      "methods" : [
+        {
+          "name" : "<init>",
+          "parameterTypes" : [
+            "java.lang.String"
+          ]
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.Tracer"
+      },
+      "type" : "org.apache.commons.logging.impl.Log4JLogger"
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.Tracer"
+      },
+      "type" : "org.apache.commons.logging.impl.LogFactoryImpl",
+      "methods" : [
+        {
+          "name" : "<init>",
+          "parameterTypes" : [ ]
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.Tracer"
+      },
+      "type" : "org.apache.commons.logging.impl.WeakHashtable",
+      "methods" : [
+        {
+          "name" : "<init>",
+          "parameterTypes" : [ ]
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.Trace"
+      },
+      "type" : "sun.security.provider.NativePRNG",
+      "methods" : [
+        {
+          "name" : "<init>",
+          "parameterTypes" : [
+            "java.security.SecureRandomParameters"
+          ]
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.Tracer"
+      },
+      "type" : "sun.security.provider.NativePRNG",
+      "methods" : [
+        {
+          "name" : "<init>",
+          "parameterTypes" : [
+            "java.security.SecureRandomParameters"
+          ]
+        }
+      ]
+    }
+  ],
+  "resources" : [
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.Tracer"
+      },
+      "glob" : "META-INF/services/org.apache.commons.logging.LogFactory"
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.Tracer"
+      },
+      "glob" : "commons-logging.properties"
+    }
+  ]
+}

--- a/tests/src/org.htrace/htrace-core/3.0.4/src/test/resources/META-INF/native-image/reachability-metadata.json
+++ b/tests/src/org.htrace/htrace-core/3.0.4/src/test/resources/META-INF/native-image/reachability-metadata.json
@@ -9,6 +9,41 @@
           "org_htrace.htrace_core.Htrace_coreTest$GreetingService"
         ]
       }
+    },
+    {
+      "condition" : {
+        "typeReached" : "org.htrace.wrappers.TraceProxy$1"
+      },
+      "type" : "org_htrace.htrace_core.Htrace_coreTest$GreetingService",
+      "methods" : [
+        {
+          "name" : "greet",
+          "parameterTypes" : [
+            "java.lang.String"
+          ]
+        },
+        {
+          "name" : "sum",
+          "parameterTypes" : [
+            "int",
+            "int"
+          ]
+        }
+      ]
+    }
+  ],
+  "resources" : [
+    {
+      "condition" : {
+        "typeReached" : "org_htrace.htrace_core.Htrace_coreTest"
+      },
+      "glob" : "META-INF/services/org.assertj.core.configuration.Configuration"
+    },
+    {
+      "condition" : {
+        "typeReached" : "org_htrace.htrace_core.Htrace_coreTest"
+      },
+      "glob" : "META-INF/services/org.assertj.core.presentation.Representation"
     }
   ]
 }

